_posts/2025-09-05-anatomy-of-vllm.md

```diff
@@ -0,0 +1,992 @@
+3. Initializes DP group (e.g. using NCCL backend).
+4. Initializes the <code>EngineCore</code> with <code>MultiProcExecutor</code> (<code>TP=4</code> on 4 GPUs as described earlier).
+5. Creates a <code>ready_event</code> (<code>threading.Event</code>).
+6. Starts an input deamon thread (<code>threading.Thread</code>) running <code>process_input_sockets(…, ready_event)</code>. Similarly starts an output thread.
+7. Still in the main thread, waits on <code>ready_event</code> until all input threads across all 4 processes (spanning the 2 nodes) have completed the coordination handshake finally executing <code>ready_event.set()</code>.
+8. Once unblocked, sends a "ready" message to the frontend with metadata (e.g., <code>num_gpu_blocks</code> available in paged KV cache memory).
+9. The main, input, and output threads then enter their respective busy loops.
+
+TL;DR: We end up with 4 child processes (one per DP replica), each running a main, input, and output thread. They complete a coordination handshake with the DP coordinator and frontend, then all three threads per process run in steady-state busy loops.
+
+<p align="center">
+<picture>
+<img src="/assets/figures/2025-vllm-anatomy/dpenginecoreproc.png" width="100%">
+</picture><br>
+<b>Figure 15</b>: distributed system with 4 DP replicas running 4 DPEngineCoreProc
+</p>
+
+<b>Current steady state</b>:
+
+* <b>Input thread</b> — blocks on the input socket until a request is routed from the API server; upon receipt, it decodes the payload, enqueues a work item via <code>input_queue.put_nowait(...)</code>, and returns to blocking on the socket.
+* <b>Main thread</b> — wakes on <code>input_queue.get(...)</code>, feeds the request to the engine; <code>MultiProcExecutor</code> runs the forward pass and enqueues results to <code>output_queue</code>.
+* <b>Output thread</b> — wakes on <code>output_queue.get(...)</code>, sends the result back to the API server, then resumes blocking.
+
+<b>Additional mechanics</b>:
+
+* <b>DP wave counter</b> — the system tracks "waves"; when all engines become idle they quiesce, and the counter increments when new work arrives (useful for coordination/metrics).
+* <b>Control messages</b> — the API server can send more than just inference requests (e.g., aborts and utility/control RPCs).
+* <b>Dummy steps for lockstep</b> — if any DP replica has work, all replicas execute a forward step; replicas without requests perform a dummy step to participate in required synchronization points (avoids blocking the active replica).
+
+> [!NOTE]
+> Lockstep clarification: this is actually only required for MoE models where the expert layers form an EP or TP group while attention layers are still DP. It's currently always done with DP - this is just because there's limited use for "built-in" non-MoE DP since you could just run multiple independent vLLMs and load-balance between them in a normal way.
```

Comment:
although there's clarification here, I'm afraid that this can be misleading. I already hear many complaints from users that using the `-dp` argument is much slower than spinning up independent vLLM instances. this is expected due to extra per-step synchronization, but not intended. vLLM's dp is only meant for MoE models with EP and DP > 1. Using dp in such cases where vLLM instances can be totally independent, is very dangerous (slow).

@njhill do you think we should raise error if `-dp` is larger than 1 while people do not enable expert parallel?

Reply:
Nick was the one who gave me this clarification during his review before i published